### PR TITLE
Store pinned static fields in the Pinned Object Heap.

### LINKED ISSUE #66043: Store static fields with `[FixedAddressValueType]` in the Pinned Object Heap.
The assembly unloadability document says that [static fields with `[FixedAddressValueType]` are incompatible with unloadability](https://github.com/dotnet/runtime/blob/main/docs/design/features/unloadability.md#fixedaddressvaluetypeattribute-for-fields-in-collectible-types) because the pin is implemented using a strong pinned handle. However, since the .NET Core 3.0 era when unloadability was introduced, the runtime has some additional gadgets up its sleeves, like the Pinned Object Heap.

The problem of unloadability with these fields could be solved by just placing them in the POH. It might also be a good idea to implement `[FixedAddressValueType]` this way _always_, to lessen the pressure of stray pinned objects in the generational heap.

I am not personally affected by this issue and I don't ask to treat it with urgency. I just got a random idea and wanted to gather some feedback. It might be decided that this attribute is so rare that working on it is not worth the effort at all.

### PR DESCRIPTION
**Pull request by teo-tsirpanis**
This PR allocates fields marked with `[FixedAddressValueType]` to the Pinned Object Heap. We introduce a new overload of `AllocateObject` that accepts a `GC_ALLOC_FLAGS`, and when allocating the static fields we pass `GC_ALLOC_PINNED_OBJECT_HEAP` if the field has that attribute.

Because we don't use a pinned handle to pin the fields anymore, the object is not permanently held in memory, which enables unloadability for assemblies with pinned static fields. All we have to do is remove the check that blocked these assemblies, and unmark the relevant unit test as incompatible with unloadability. I also removed a now-unused tracking mechanism for thread-static pinned fields.

Fixes #66043

### CODE REVIEW

File: src/coreclr/vm/gchelpers.h [hunk 1]
```diff
@@ -53,11 +53,24 @@ OBJECTREF DupArrayForCloning(BASEARRAYREF pRef);
...
                          , bool fHandleCom = true
 #endif
     );
 
+inline OBJECTREF AllocateObject(MethodTable *pMT
+#ifdef FEATURE_COMINTEROP
+                                , bool fHandleCom = true
+#endif
+    ) {
+    return AllocateObject(pMT, GC_ALLOC_NO_FLAGS
```
Comment (janvorli) [inline src/coreclr/vm/gchelpers.h:67]: A nit - please fix the indentation of this line

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Comment by janvorli** [comment]
@teo-tsirpanis you'll also need to remove the library negative test for the FixedAddressValueType attribute and unloadability - it is causing the CI failures.

File: .vscode/settings.json
Comment (teo-tsirpanis) [inline .vscode/settings.json:1]: Will remove it.

**Comment by MichalPetryka** [comment]
> Because we don't use a pinned handle to pin the fields anymore, the object is not permanently held in memory, which enables unloadability for assemblies with pinned static fields.

Worth noting is that this will unload the assembly when the user uses `Unsafe.AsPointer` on the field since nothing with keep it alive and then such access could corrupt the process.

File: src/coreclr/vm/gchelpers.h [hunk 2]
```diff
@@ -53,11 +53,24 @@ OBJECTREF DupArrayForCloning(BASEARRAYREF pRef);
...
 #ifdef FEATURE_COMINTEROP
                          , bool fHandleCom = true
 #endif
     );
 
+inline OBJECTREF AllocateObject(MethodTable *pMT
+#ifdef FEATURE_COMINTEROP
+                                , bool fHandleCom = true
+#endif
+    ) {
```
[the remark this answers is not in the document]
Comment (teo-tsirpanis) [reply src/coreclr/vm/gchelpers.h]: Better now?

**Comment by mangod9** [comment]
@teo-tsirpanis, could you please rebase the PR and then we could run the CI again to get it merged? Thanks

**Comment by teo-tsirpanis** [comment]
@mangod9 done. CI is green.

**Review by jkotas — APPROVED** [review]
Thank you!

**Pull request merged**